### ISSUE TITLE
Support teacher forcing for AspuruGuzikAutoencoder

### ISSUE THREAD

**Issue by john-judge** [deepchem/deepchem#2456]
## 🚀 Feature

Proposal to support teacher forcing in `AspuruGuzikAutoEncorder` (in `seq2seq.py`) with a teacher forcing probability parameter.

## Motivation

As a small group of students, we are interested in characterizing the performance tradeoffs in teacher forcing. While models subsequent to `chemical_vae` have effectively dealt with the syntactic invalidity of generated molecules, we plan to look at SMILES strings as a case-study of teacher forcing to attempt to learn a context-free grammar without explicit constraint. We are planning to develop this enhancement but appreciate any input or review.

## Additional context

#1026 implemented the `chemical_vae` character-based VAE from [aspuru-guzik-group/chemical_vae](https://github.com/aspuru-guzik-group/chemical_vae) without teacher forcing.

**Comment by rbharath** [comment]
Sounds like a great project! I'd recommend joining our developer calls https://forum.deepchem.io/t/joining-the-deepchem-developer-calls/354 to get feedback or assistance on your work

**Comment by john-judge** [comment]
https://github.com/molecularsets/moses/issues/7#issue-387506021

**Comment by john-judge** [comment]
These are just notes for our group project. The Colab mentioned is internal to our group, but we expect to have a pull request within the next few weeks.

## Goals
- expose KL cost annealing parameters to AspuruGuzikAutoEncoder
- implement teacher forcing in AspuruGuzikAutoEncoder
- backward-compatibility (including making any new dependencies soft where possible)
- characterize training methods including: teacher forcing, KL cost annealing [, Batch Norm?, teacher forcing annealing?]

## Teacher Forcing 
### Strategy/design:

We would want to enable this to teacher-force during training but free-sample during generation.   The[ way that Aspuru-Guzik-group did this](http://tensorflow.biotecan.com/python/Python_1.8/tensorflow.google.cn/api_docs/python/tf/keras/backend/in_train_phase.html) was to use the keras function `backend.in_train_phase` because the training flag can act as the switch for which behavior to use. 

### Starting point
Classes copied in from `deepchem`:
 - `AspuruGuzikAutoEncoder`
 - `SeqToSeq`
 - `VariationalRandomizer`

Copied from original Aspuru-Guzik-group (`chemvae`):
 - `TerminalGRU`
 - `sampled_rnn (function)`

### Progress in TerminalGRU:
- Now supports probabilistic teacher forcing (original had fixed `teacher_forcing_ratio = 1`)
- `TerminalGRU` keras version issues:
-- Note that `preprocess_input` is missing as a method of `TerminalGRU` (it should inherit from `keras.layers.recurrent.GRU`, but issue probably caused by the fact that `TerminalGRU` uses an older version of keras than deepchem). It's just an identity function for GPUs and I believe its CPU-specific logic is performance-only. For now, let's remove calls to this method or otherwise replace with identity function.
-- In `TerminalGRU` calls to 'add_weight' need to have the first argument passed as a named parameter rather than positionally. Put in "shape=" for the first argument.


### Progress in `AspuruGuzikAutoEncoder`:
 - `AspuruGuzikAutoEncoder` now supports parameter `teacher_forcing_ratio`, which it later passes to the correct parameter of a `TerminalGRU` instantiation as the third and final GRU layer. Defaults to value 0.0.
 - `_create_decoder()` has only 2 normal GRU layers instead of 3
 - the call to the initialized `TerminalGRU` passes the input as `[prev_layer, input]`, where 'input' should be the true input. However, this is where I'm probably incorrect. Compare this to how the original implementation passes this in as `[x_dec, true_seq_in]` in the [original Aspuru-Guzik](https://github.com/aspuru-guzik-group/chemical_vae/blob/master/chemvae/models.py) -- they read in the latent space as one Input, and the `true_seq_in` (ground truth) as another `Input`. I believe we will need to pass ground truth input into `_create_decoder()` and update calls to it, which means overriding much of `AspuruGuzikAutoEncoder`'s call to its superclass `SeqToSeq`'s init.

### To do:
- next step mentioned above, also continue in teacher forcing attempt Colab 
- check input/output shapes
- train and compare to our previous results

**Comment by rbharath** [comment]
Looks like some cool features! As a quick note make sure to read our coding conventions: https://deepchem.readthedocs.io/en/latest/development_guide/coding.html

DeepChem reviews can be involved so I'd suggest submitting a draft PR for us to review. If the changes are large, it might have to be broken into multiple pull requests to keep things sane. I'd also highly encourage your group to come to office hours: https://forum.deepchem.io/t/announcing-the-deepchem-office-hours/293

I can give feedback and suggestions there

**Comment by john-judge** [comment]
Thank you, @rbharath ! I'll stop by on Friday. I think my group may want to put KL cost annealing through as a separate PR so that we can quickly get a sense of the DeepChem development process. It's a much smaller change but likely should still have its own benchmarking results.

**Comment by john-judge** [comment]
We ended up using the MOSES generative molecular repository to investigate teacher-forced Aspuru-Guzik VAE.
We also followed @rbharath 's suggestion with SELFIES, using the Aspuru-Guzik's `selfies` Python library and combining it with DeepChem's `AspuruGuzikAutoEncoder` (mapping SELFIES tokens to single-character tokens) and found 100% validity for generated molecules (as validated by RDKit's SMILES evaluator)

This points us in the direction of branching out to metrics beyond validity (e.g. drug-likeness, synthesizability) to more precisely characterize how chemically rich this SELFIES-based latent space is, and comparing it to the teacher-forced SMILES-based latent space in terms of utility for drug discovery / optimization.

Closing this issue
